src/Microsoft.AspNet.OData.Shared/Builder/DerivedTypeConstraintSet.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using Microsoft.AspNet.OData.Common;
+using Microsoft.OData.UriParser;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    internal class DerivedTypeConstraintSet : HashSet<Type>
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public DerivedTypeConstraintSet()
+        {
+        }
+
+        public DerivedTypeConstraintSet(Type baseType)
+        {
+            ClrBaseType = baseType;
+        }
+
+        public Type ClrBaseType { get; set; }
+
+        public void ValidateAndAddConstraints(IEnumerable<Type> types)
+        {
+            if (ClrBaseType == null)
+            {
+                Error.InvalidOperation(SRResources.BaseTypeNotSpecified);
+            }
+
+            ValidateAndAddConstraints(ClrBaseType, types);
+        }
+
+        public void ValidateAndAddConstraints(Type baseType, IEnumerable<Type> types)
+        {
+            if (!types.Any())
+            {
+                ValidateAndAddSingleConstraint(baseType, baseType);
```

Comment:
>ValidateAndAddSingleConstraint [](start = 16, length = 30)

What if other derived types have already been added to the constraint? What's the behavior of *also* adding the base type? What if other derived types are added *after* adding the base type?